### PR TITLE
Allow replacing fetch tile and decode image for local Dem manager

### PR DESCRIPTION
**Pull request by HarelM**
I've added the ability to replace decode image and fetch tile.
I've also moved some stuff around and did some minor renaming.
Let me know what you think.

This can facilitate for #212 for the case of local dem manager, but not for remote dem manager.

### CODE REVIEW

**Comment by HarelM** [comment]
CC: @acalcutt

**Comment by HarelM** [comment]
The following script seems to work as expected I think, I tried to use the exported parts of this library
```ts
import { writeFileSync } from "fs";
import type { DemTile, Encoding } from "./dist/types";
import exported from "./dist/index";
import { PNG } from "pngjs";

async function decodeImageNode(
    blob: Blob,
    encoding: Encoding,
    abortController: AbortController,
  ): Promise<DemTile> {
    const buffer = await blob.arrayBuffer();
    const png = PNG.sync.read(Buffer.from(buffer));
    const parsed = exported.decodeParsedImage(png.width, png.height, encoding, png.data as any as Uint8ClampedArray);
    if (Boolean(abortController?.signal?.aborted)) return null as any as DemTile;
    return parsed;
  }

let manager = new exported.LocalDemManager({
    demUrlPattern: "https://www.example.com/TerrainRGB/{z}/{x}/{y}.png", // This is the URL of the DEM tiles
    cacheSize: 100,
    encoding: "mapbox",
    maxzoom: 12,
    timeoutMs: 10000,
    decodeImage: decodeImageNode,
});

manager.fetchContourTile(12,2446,1655, {levels: [10]}, new AbortController()).then((tile) => {
    writeFileSync("12-2446-1655.mvt", Buffer.from(tile.arrayBuffer));
});
```

File: src/local-dem-manager.ts [hunk 1]
```diff
@@ -92,24 +75,10 @@ export class LocalDemManager implements DemManager {
     return this.tileCache.get(
       url,
       (_, childAbortController) => {
-        const options: RequestInit = {
-          signal: childAbortController.signal,
-        };
         timer?.fetchTile(url);
-        const mark = timer?.marker("fetch");
```
Comment (msbarry) [inline src/local-dem-manager.ts:99]: This metrics gathering is missing from the getTile implementation, can we add it back in outside of the getTile call, something like:

```ts
const mark = timer?.mark("fetch");
return withTimeout(
  this.timeoutMs,
  this.getTile(url, childAbortController).finally(() => mark?.()),
  childAbortController,
);
```

**Review by msbarry — COMMENTED** [review]
Looks good overall! A couple of small suggestions, but can you also run `npm run format` and add a test with a local DEM manager with custom function for fetch and resolve?

File: src/types.ts [hunk 2]
```diff
@@ -114,3 +110,55 @@ export interface Timing {
...
+
+export type DemManagerInitizlizationParameters = DemManagerRequiredInitizlizationParameters & {
+  decodeImage?: DecodeImageFunction;
+  getTile?: GetTileFunction;
+  actor?: Actor<WorkerDispatch>;
+}
+
+export type InitMessage = DemManagerRequiredInitizlizationParameters & {
+  managerId: number;
+}
```
[the remark this answers is not in the document]
Comment (HarelM) [reply src/types.ts]: It already does, the required fields do not contain the added functions.

File: src/local-dem-manager.ts [hunk 3]
```diff
@@ -1,46 +1,36 @@
...
-    y: number,
-    options: IndividualContourTileOptions,
-    abortController: AbortController,
-    timer?: Timer,
-  ): Promise<ContourTile>;
+const defaultGetTile: GetTileFunction = async (url: string, abortController: AbortController) => {
+  const options: RequestInit = {
+    signal: abortController.signal,
+  };
+  return fetch(url, options).then(async (response) => {
```
[the remark this answers is not in the document]
Comment (HarelM) [reply src/local-dem-manager.ts]: Done in a93384f9db394851c1ba142e970f8406e18d067e

File: src/local-dem-manager.ts (hunk 1 above)
Comment (HarelM) [reply src/local-dem-manager.ts:99]: Sure, wasn't entirely sure how to solve this...

Comment (HarelM) [reply src/local-dem-manager.ts:99]: Done in edfe7d1

**Comment by HarelM** [comment]
I've done the formatting and added a e2e test, although it might be better to move it to "local-dem-manager.test.ts" as a unit test for that class, not sure, let me know what you prefer...

File: src/local-dem-manager.ts [hunk 4]
```diff
@@ -55,26 +47,19 @@ export class LocalDemManager implements DemManager {
...
-    timeoutMs: number,
-  ) {
-    this.tileCache = new AsyncCache(cacheSize);
-    this.parsedCache = new AsyncCache(cacheSize);
-    this.contourCache = new AsyncCache(cacheSize);
-    this.timeoutMs = timeoutMs;
-    this.demUrlPattern = demUrlPattern;
-    this.encoding = encoding;
-    this.maxzoom = maxzoom;
+  constructor(options: DemManagerInitizlizationParameters) {
```
Comment (msbarry) [inline src/local-dem-manager.ts:53]: Looks like this change broke the benchmark script. Can you update the LocalDemManager constructor call there? You can test using `npm run build` and `node benchmark.js`

File: src/types.ts [hunk 5]
```diff
@@ -114,3 +109,63 @@ export interface Timing {
...
+};
+
+export type DemManagerInitizlizationParameters =
+  DemManagerRequiredInitizlizationParameters & {
+    decodeImage?: DecodeImageFunction;
+    getTile?: GetTileFunction;
+    actor?: Actor<WorkerDispatch>;
+  };
+
+export type InitMessage = DemManagerRequiredInitizlizationParameters & {
```
Comment (msbarry) [inline src/types.ts:169]: typo: `Initialization`

Comment (HarelM) [reply src/types.ts:169]: Fixed. Thanks.

File: src/local-dem-manager.ts (hunk 4 above)
Comment (HarelM) [reply src/local-dem-manager.ts:53]: Should be fixed now. I would recommend changing the settings for CI run to allow veteran github users to activate the CI.

**Review by msbarry — APPROVED** [review]

**Comment by msbarry** [comment]
Thank you! Do you think it would make sense to upstream a top-level node.js-ready entrypoint using pngjs/possibly pmtiles?

**Comment by HarelM** [comment]
I think @acalcutt is working on a script that can allow running this using a shell script to be added to the bin folder of this package.
Note that the constructor arguments change is a breaking change in terms of public API so you might need to release a major version.

**Comment by acalcutt** [comment]
I made this code to get all the tiles under a specific parent tile which is based of similar code @HarelM had given me
https://github.com/acalcutt/maplibre-contour-pmtiles/blob/main/src/generate-countour-tile-batch.ts
`Usage: npx tsx ./src/generate-countour-tile-batch.ts --x <x> --y <y> --z <z> --sFile <sFile> --sEncoding <sEncoding> --sMaxZoom <sMaxZoom> --increment <increment> --oMaxZoom <oMaxZoom> --oDir <oDir>"`

and this batch script the gets all tiles at a specific level and runs the above script for each
https://github.com/acalcutt/maplibre-contour-pmtiles/blob/main/scripts/generate_contour_tiles.sh
```
Usage: ./generate_contour_tiles.sh --sFile <path> --oDir <path> [options]
 Options:
  --increment <value> Increment value (default: 10)
  --sMaxZoom <value> Source Max Zoom (default: 8)
  --sEncoding <encoding> Source Encoding (default: mapbox) (must be 'mapbox' or 'terrarium')
  --sFile <path>  TerrainRGB or Terrarium PMTiles File Path or URL (REQUIRED)
  --oDir <path>  Output Directory (REQUIRED)
  --oMaxZoom <value> Output Max Zoom (default: 8)
  --oMinZoom <value> Output Min Zoom (default: 5)
  -v|--verbose  Enable verbose output
  -h|--help  Show this usage statement
```

I used it to make this z5-z11 contour at 10m 
increments https://tiles.wifidb.net/data/JAXA_AW3D30_2024_contour_z5-Z11_vector/#5/0/0
(though I am making a new one z5-z12 with slighly tower increments for the lower soom levels)

It shouldn't be hard to move that to this new method of fetch tile method I wouldn't think.

**Comment by acalcutt** [comment]
Would it be possible to override 'fetchTile' in the same way? as decodeImage?

**Comment by HarelM** [comment]
It is part of this PR, yes.

**Comment by acalcutt** [comment]
ah, nice. i think i see in the tests how it would be used

**Comment by acalcutt** [comment]
I was trying to use your maplibre-contour changes with this code,
https://github.com/WifiDB/maplibre-contour/blob/ea60d710750c389ed05c0443a0873b18b1373faa/scripts/generate-countour-tile-pyramid.ts
but it keeps complaining about
```
SyntaxError: The requested module '../dist/index' does not provide an export named 'default'
    at ModuleJob._instantiate (node:internal/modules/esm/module_job:132:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:214:5)
    at async ModuleLoader.import (node:internal/modules/esm/loader:329:24)
    at async loadESM (node:internal/process/esm_loader:34:7)
    at async handleMainPromise (node:internal/modules/run_main:113:12)
```
but when i look at those index files the do seem like they have a default export... I did try to delete dist and recreated but it didn't help


The other thing is I noticed $url only seems to be the only parameter for getTile,. It would be nice if zxy parameters where given to so it doesn't need to be split out of the url

**Comment by HarelM** [comment]
Did you make sure to rebuild dist folder?
I think I managed to get things working as to what I wrote in the comment here:
https://github.com/onthegomap/maplibre-contour/pull/359#issuecomment-2551709029

I'll take a look later tonight to see if I can make it work. 
@msbarry any chance you can publish the new version to npm?

**Comment by HarelM** [comment]
I've updated the above comment to something that is working for me, I'm not sure what's not well defined in terms of typescript, but it can run using tsx, vscode doesn't like what I wrote though...

**Comment by acalcutt** [comment]
Nice, I was able to get my local pmtiles script to work with this change

I modified the code I posted  above that builds all tiles under a parent tile to work with this new feature
https://github.com/WifiDB/maplibre-contour/blob/fetch_test/scripts/generate-countour-tile-pyramid.ts

I also put the node/pmtiles functions I had made into it's own file that can be referenced
https://github.com/WifiDB/maplibre-contour/blob/fetch_test/scripts/node-pmtiles-adapter.ts

In 'generate-countour-tile-pyramid.ts' I added a switch if demUrl starts with pmtiles:// so it should be able to either work with pmtiles or a regular http(s) url https://github.com/WifiDB/maplibre-contour/blob/fetch_test/scripts/generate-countour-tile-pyramid.ts#L218-L240, 

it should start with something like this for pmtiles
`npx tsx generate-countour-tile-pyramid.ts --x 17 --y 6 --z 5 --demUrl pmtiles:///path/to/dem.pmtiles --sEncoding terrarium --sMaxZoom 8 --increment 10 --oMaxZoom 8 --oDir ./out`
or
`npx tsx generate-countour-tile-pyramid.ts --x 17 --y 6 --z 5 --demUrl pmtiles://https://url/path/to/dem.pmtiles --sEncoding terrarium --sMaxZoom 8 --increment 10 --oMaxZoom 8 --oDir ./out`

or for a tile url
`npx tsx generate-countour-tile-pyramid.ts --x 17 --y 6 --z 5 --demUrl https://url/of/dem/source/{z}/{x}/{y}.png --sEncoding terrarium --sMaxZoom 8 --increment 10 --oMaxZoom 8 --oDir ./out`

**Comment by msbarry** [comment]
OK I just published a 0.1.0 release: https://www.npmjs.com/package/maplibre-contour/v/0.1.0

**Comment by HarelM** [comment]
Thanks @msbarry! I was able to create a new project (external to maplibre-contour) and use the following code to generate a tile locally.
Note that the types are used from the `dist` folder which can create a breaking change if you change internal implementation.
I would advise to use something like `dts-bundle-generator` or an equivalent to bundle all the types into a single `index.d.ts` file so that everything is exported from the same path (I wish typescript folks would implement that already and avoid the need to write it in every package I maintain, but that's a different story).
```ts
import { writeFileSync } from "fs";
import mlcontour from "maplibre-contour";
import { DemTile, Encoding } from "maplibre-contour/dist/types";
import { PNG } from "pngjs";

async function decodeImageNode(
    blob: Blob,
    encoding: Encoding,
    abortController: AbortController,
  ): Promise<DemTile> {
    const buffer = await blob.arrayBuffer();
    const png = PNG.sync.read(Buffer.from(buffer));
    const parsed = mlcontour.decodeParsedImage(png.width, png.height, encoding, png.data as any as Uint8ClampedArray);
    if (Boolean(abortController?.signal?.aborted)) return null as unknown as DemTile;
    return parsed;
  }

let manager = new mlcontour.LocalDemManager({
    demUrlPattern: "https://www.expample.com/{z}/{x}/{y}.png", // This is the URL of the DEM tiles
    cacheSize: 100,
    encoding: "mapbox",
    maxzoom: 12,
    timeoutMs: 10000,
    decodeImage: decodeImageNode,
});

manager.fetchContourTile(12,2446,1655, {levels: [10]}, new AbortController()).then((tile) => {
    writeFileSync("12-2446-1655.mvt", Buffer.from(tile.arrayBuffer));
});
```

**Pull request merged**